extern/test_bmi_cpp/src/test_bmi_cpp.cpp

```diff
@@ -0,0 +1,429 @@
+  }
+  iter = std::find(this->input_var_names.begin(), this->input_var_names.end(), name);
+  if(iter != this->input_var_names.end()){
+    return this->input_var_locations[iter - this->input_var_names.begin()];
+  }
+  throw new std::runtime_error("GetVarLocation called for non-existent variable: "+name);
+}
+
+int TestBmiCpp::GetVarNbytes(std::string name){
+  int item_size = this->GetVarItemsize(name);
+
+  // this will never actually get used, but mimicing the C version...
+  // if a different item_count isn't found below, the call above will have already thrown.
+  int item_count = this->num_time_steps; 
+
+  auto iter = std::find(this->output_var_names.begin(), this->output_var_names.end(), name);
+  if(iter != this->output_var_names.end()){
+    item_count = this->output_var_item_count[iter - this->output_var_names.begin()];
+  }
+  iter = std::find(this->input_var_names.begin(), this->input_var_names.end(), name);
+  if(iter != this->input_var_names.end()){
+    item_count = this->input_var_item_count[iter - this->input_var_names.begin()];
+  }
+  return item_size * item_count;
+}
+
+std::string TestBmiCpp::GetVarType(std::string name){
+  auto iter = std::find(this->output_var_names.begin(), this->output_var_names.end(), name);
+  if(iter != this->output_var_names.end()){
+    return this->output_var_types[iter - this->output_var_names.begin()];
+  }
+  iter = std::find(this->input_var_names.begin(), this->input_var_names.end(), name);
+  if(iter != this->input_var_names.end()){
+    return this->input_var_types[iter - this->input_var_names.begin()];
+  }
+  throw new std::runtime_error("GetVarType called for non-existent variable: "+name);
+}
+
+std::string TestBmiCpp::GetVarUnits(std::string name){
+  auto iter = std::find(this->output_var_names.begin(), this->output_var_names.end(), name);
+  if(iter != this->output_var_names.end()){
+    return this->output_var_units[iter - this->output_var_names.begin()];
+  }
+  iter = std::find(this->input_var_names.begin(), this->input_var_names.end(), name);
+  if(iter != this->input_var_names.end()){
+    return this->input_var_units[iter - this->input_var_names.begin()];
+  }
+  throw new std::runtime_error("GetVarUnits called for non-existent variable: "+name);
+}
+
+void TestBmiCpp::Initialize(std::string file){
+  if (file == "")
+    throw new std::runtime_error("No configuration file path provided.");
+
+  this->read_init_config(file);
+
+  this->current_model_time = this->GetStartTime();
+
+  // If neither of these is read from config (remain 0 as initialized), fall back to default for ts count
+  if (this->num_time_steps == 0 && this->model_end_time == 0) {
+      this->num_time_steps = DEFAULT_TIME_STEP_COUNT;
+  }
+  // Now at least one must be set
+  if(!(this->model_end_time != 0 || this->num_time_steps != 0)){
+    throw new std::runtime_error("Configuration did not specify sufficient time range information.");
```

Comment:
I misinterpreted this condition--this is more of an `assert` type situation because it shouldn't be possible on this path, or at least not unless `DEFAULT_TIME_STEP_COUNT` is zero.